src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs

```diff
@@ -188,6 +189,132 @@ public partial class JsonTypeInfo<T>
             }
         }
 
+        internal async Task SerializeAsync(
+            PipeWriter utf8Json,
+            T? rootValue,
+            CancellationToken cancellationToken,
+            object? rootValueBoxed = null)
+        {
+            Debug.Assert(IsConfigured);
+            Debug.Assert(rootValueBoxed is null || rootValueBoxed is T);
+
+            if (CanUseSerializeHandlerInStreaming)
+            {
+                // Short-circuit calls into SerializeHandler, if the `CanUseSerializeHandlerInStreaming` heuristic allows it.
+                Debug.Assert(SerializeHandler != null);
+                Debug.Assert(CanUseSerializeHandler);
+                Debug.Assert(Converter is JsonMetadataServicesConverter<T>);
+
+                Utf8JsonWriter writer = Utf8JsonWriterCache.RentWriter(Options, utf8Json);
+
+                try
+                {
+                    SerializeHandler(writer, rootValue!);
+                    writer.Flush();
+                }
+                finally
+                {
+                    // Record the serialization size in both successful and failed operations,
+                    // since we want to immediately opt out of the fast path if it exceeds the threshold.
+                    OnRootLevelAsyncSerializationCompleted(writer.BytesCommitted + writer.BytesPending);
+
+                    Utf8JsonWriterCache.ReturnWriter(writer);
+                }
+
+                await utf8Json.FlushAsync(cancellationToken).ConfigureAwait(false);
+            }
+            else if (
+#if NETCOREAPP
+                !typeof(T).IsValueType &&
+#endif
+                Converter.CanBePolymorphic &&
+                rootValue is not null &&
+                Options.TryGetPolymorphicTypeInfoForRootType(rootValue, out JsonTypeInfo? derivedTypeInfo))
+            {
+                Debug.Assert(typeof(T) == typeof(object));
+                await derivedTypeInfo.SerializeAsObjectAsync(utf8Json, rootValue, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                bool isFinalBlock;
+                WriteStack state = default;
+                state.Initialize(this,
+                    rootValueBoxed,
+                    supportContinuation: true,
+                    supportAsync: true);
+
+                state.CancellationToken = cancellationToken;
+
+                using var writer = new Utf8JsonWriter(utf8Json, Options.GetWriterOptions());
+
+                try
+                {
+                    do
+                    {
+                        // Arbitrary threshold, default Pipe buffers are 4k so this allows 4 buffers to be created and filled (mostly) before flushing.
+                        state.FlushThreshold = (int)(16384 * JsonSerializer.FlushThreshold);
```

Comment:
In the `Stream` serializer this threshold is informed by the capacity of the buffer writer, which in turn is initialized by the user-defined `JsonSerializerOptions.DefaultBufferSize`. Why can't we use the same approach here?

Reply:
> Presumably capacity information is not available in PipeWriter?

Correct. The initial spike of Pipes+Json back in the day used a trick where it calls `GetMemory()` and looks at the length of the returned buffer. This sort of works except if the pipe has some data already written to it then the returned buffer might be smaller than the normal buffer size.

> 16384 happens to be the default value of `JsonSerializerOptions.DefaultBufferSize` so why not just use that?

Yeah this works, lets the user influence the size and is the same "arbitrary" value I chose.

Reply:
I'm wondering what the "right" way to do this should be. In the `Stream` overloads `DefaultBufferSize` acts as the seed value for a growable buffer that the serializer owns, but here we're fixing it on the understanding that the `PipeWriter` internals (currently) match our own defaults. This means that setting `DefaultBufferSize` to 1 will have very different performance characteristics in `Stream` or `IAsyncEnumerable` compared to this.

I think it would make sense to expose a `Capacity` property or something similar on `PipeWriter` so that the serializer can better cooperate with it. @davidfowl thoughts?

Reply:
I think that can be a follow up. Doesn't need to be a part of this PR.

Reply:
In that case, let's not have this depend on `DefaultBufferSize`. But at the same time, we needn't hardcode the value either:

```suggestion
                        state.FlushThreshold = (int)((4 * PipeOptions.Default.MinimumSegmentSize) * JsonSerializer.FlushThreshold);
```